File: Behavioral/strategy.java
public class Strategy {
    public interface SoundBehavior {
        public String makeSound();
    }

    public static class MeowSound implements SoundBehavior {
        public String makeSound() {
            return "Meow meow meow!";
        }
    }

    public static class PurrSound implements SoundBehavior {
        public String makeSound() {
            return "Purrr... Purrrrr...";
        }
    }

    public static abstract class Cat {
        SoundBehavior soundBehavior;
        String name;

        public Cat(String name) {
            this.name = name;
        }

        public void performSound() {
            String sound = String.format("%s: %s", name, soundBehavior.makeSound());
            System.out.println(sound);
        }

        public void setSoundBehavior(SoundBehavior ab) {
            soundBehavior = ab;
        }
    }

    public static class Siamese extends Cat {
        public Siamese(String name) {
            super(name);
            soundBehavior = new MeowSound();
        }
    }

    public static class MaineCoon extends Cat {
        public MaineCoon(String name) {
            super(name);
            soundBehavior = new PurrSound();
        }
    }

    public static void main(String[] args) {
        Cat siamese = new Siamese("Simba");
        Cat maineCoon = new MaineCoon("Oscar");

        siamese.performSound();
        maineCoon.performSound();
        siamese.setSoundBehavior(new PurrSound());
        siamese.performSound();
    }
}

File: Behavioral/strategy.ts
interface SoundBehavior {
    makeSound(): void;
}

class MeowSound implements SoundBehavior {
    makeSound() {
        return 'Meow meow meow!';
    }
}

class PurrSound implements SoundBehavior {
    makeSound() {
        return 'Purrr... Purrrrr...';
    }
}

abstract class Cat {
    protected _soundBehavior: SoundBehavior;
    protected _name: string;

    constructor(name: string) {
        this._name = name;
    }

    public performSound() {
        console.log(`${this._name}: ${this._soundBehavior.makeSound()}`);
    }

    public setSoundBehavior(sb: SoundBehavior) {
        this._soundBehavior = sb;
    }
}

class Siamese extends Cat {
    constructor(name) {
        super(name);
        this._soundBehavior = new MeowSound();
    }
}

class MaineCoon extends Cat {
    constructor(name) {
        super(name);
        this._soundBehavior = new PurrSound();
    }
}

let siamese = new Siamese('Simba'),
    maineCoon = new MaineCoon('Oscar');

siamese.performSound();
maineCoon.performSound();
siamese.setSoundBehavior(new PurrSound());
siamese.performSound();
